torch/utils/_foreach_utils.py

```diff
@@ -39,7 +39,9 @@ def _group_tensors_by_device_and_dtype(
     }
 
 
+def _device_has_foreach_support(device: torch.device) -> bool:
+    return device.type in (_get_foreach_kernels_supported_devices() + ["cpu"]) and not torch.jit.is_scripting()
+
+
 def _has_foreach_support(tensors: List[Tensor], device: torch.device) -> bool:
-    if device.type not in set(_get_foreach_kernels_supported_devices() + ["cpu"]) or torch.jit.is_scripting():
-        return False
-    return all(t is None or type(t) == torch.Tensor for t in tensors)
+    return _device_has_foreach_support(device) and all(t is None or type(t) == torch.Tensor for t in tensors)
```

Comment:
Nice, a good followup is to relax the type here to allow for nn.Parameter at the very least. Would be an easy PR 😛

Reply:
I think we have to be careful for `nn.Parameter` since we need to know its underlying type (under the `nn.Parameter`). Do you know what is the canonical way to do that?